### ISSUE TITLE
Temporary directory does not get removed

### ISSUE THREAD

**Issue by braggpeaks** [Kozea/WeasyPrint#2090]
weasyprint apparently creates a temporary directory but does not remove the directory afterwards.

Windows folder location: C:\Users\daniel.kellner\AppData\Local\Temp
Subfolder naming example: weasyprint-0qjsm58z

The folders are empty but I believe they should be removed to not leave unnecessary folders (I currently have 8000 of these weasyprint-* folders, don't ask 😂).

**Comment by liZe** [comment]
Hi!

The folders is supposed to be removed [here](https://github.com/Kozea/WeasyPrint/blob/4d93769a06e2a5a9d158daf2b825bbcff143ce4f/weasyprint/text/fonts.py#L256-L258), but:
- It is not guaranteed that `__del__()` methods are called for objects that still exist when the interpreter exits (quoting the Python documentation), and
- Windows likes to keep references to folders/files even when they’re not used anymore, preventing them to be deleted.

The "real" solution would be to only use data in memory (fonts are not that big). It should be possible to keep config files in memory thanks to [`FcConfigParseAndLoadFromMemory`](https://fontconfig.pages.freedesktop.org/fontconfig/fontconfig-devel/fcconfigparseandloadfrommemory.html) but there’s nothing to replace `FcConfigAppFontAddFile`.

If someone finds a better solution, I’ll be happy to try!

**Comment by seidnerj** [comment]
This is pretty annoying, it is cluttering the /tmp folder. This could also be the case if there was an errors when calling rmtree (as ignore_errors is set tot True).

**Comment by liZe** [comment]
> This is pretty annoying, it is cluttering the /tmp folder.

We now use `FcConfigParseAndLoadFromMemory`, and I’ve just added a commit that avoids creating useless temporary folders. It means that you’ll only have the folders if you use `@font-face` rules, but the problem is not solved.

Unfortunately, there’s no function in Fontconfig to add fonts from memory (see almost [11-year old Fontconfig issue](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/issues/12)). Without such a function, as explained above, I didn’t find a reliable solution, ideas are welcome.

> This could also be the case if there was an errors when calling rmtree (as ignore_errors is set tot True).

As far as I can remember, Windows was randomly trying hard to keep references to this folder, preventing `rmtree` from removing the folder.

Again, the real solution would be to avoid temporary folders, and it would require a new function in Fontconfig.

**Comment by seidnerj** [comment]
I understand. Well, if this will only happen when using @font-face then that solves this for me at least 🙂

**Comment by max-blau** [comment]
@liZe Hey, I'm running into this issue when using font-face rules.
I understand theres a limitation coming from `Fontconfig` that requires the use of a file vs memory for adding fonts. But is there any way to refactor how the tmp files are created so that they're automatically cleaned up in certain workflows? Looking through the source, this seems to be difficult - so alternatively, would it be possible to surface an option in `FontConfiguration` or somewhere else for callers to specify a tmp directory for `weasyprint` to use? That way we can guarantee its cleaned up after we're done generating a PDF.
It's not the most elegant solution, but it would at least give some control

FYI its technically possible to do this now, but is also kind of hacky:
```
with tempfile.TemporaryDirectory(prefix='my-tmp-dir-') as tmpdir:
    font_config = FontConfiguration()
    # remove the existing tmp folder if it already exists
    # required on older version of weasyprint, don't think its required for latest (67.0)
    if font_config._folder is not None:
        shutil.rmtree(font_config._folder)
    font_config._folder = Path(tmpdir)
```

**Comment by liZe** [comment]
> Looking through the source, this seems to be difficult

Yes, it is, because we don’t know when the font configuration will never be used again.

> so alternatively, would it be possible to surface an option in `FontConfiguration` or somewhere else for callers to specify a tmp directory for `weasyprint` to use?

We already have the `._folder` attribute, that’s not public (because we don’t want users to rely on it) but that can be used at your own risk 😄. I don’t think that we want a public attribute or API for that, because it should ideally be only an implementation detail, but… maybe we can do something.

Adding a public method that removes the folder (i.e. that does what `__del__` currently does) would give the users the possibility to remove it manually. We could also call this method everywhere we create a font config internally. Of course, we could also keep `__del__` to delete the folder implicitly, as it is done currently.

@max-blau What do you think about that?

**Comment by max-blau** [comment]
@liZe that sounds great - for my use case, I instantiate the FontConfiguration once, set up some CSS, generate a PDF, then am done. So being able to manually remove the folder works perfectly.

**Comment by liZe** [comment]
@max-blau Cool! Would you be interested in opening a pull request, or just to test?

**Comment by max-blau** [comment]
@liZe I can open a PR - if its ok though, I'll probably just add a method and then leave it up to the main contributors to use it in other places if required
